femtools/Time_Period.F90

```diff
@@ -0,0 +1,205 @@
+
+module time_period
+  use FLDebug
+  use global_parameters, only: OPTION_PATH_LEN
+  use embed_python
+  use spud
+  use parallel_tools
+  use timers
+
+  implicit none
+
+  !! Type for holding information about a previous dump period
+  type time_period_type
+    private
+
+    !! whether the previous dump times have been initialised
+    logical :: last_times_initialised = .false.
+
+    !! the requested time between dumps in model seconds
+    real :: real_dump_period = 0.0
+    !! the requested time between dumps in timesteps
+    integer :: int_dump_period = 0
+    !! the last model time at which a dump was performed (or when this
+    !! control struct was initialised)
+    real :: last_dump_time
+
+    !! the last cpu time at which a dump was performed
+    real :: last_dump_cpu_time
+    !! the last wall time at which a dump was performed
+    real :: last_dump_wall_time
+  end type time_period_type
+
+contains
+  function should_output(CS, current_time, timestep, option_path)
+    !!< Given a time period control struct, should a new dump be
+    !!< created at this moment in time (considering model time, the
+    !!< current timestep, CPU time, and walltime)?
+
+    logical :: should_output
+
+    type(time_period_type), intent(inout) :: CS
+    real, intent(in) :: current_time
+    integer, intent(in) :: timestep
+
+    character(len=*), intent(in) :: option_path
+
+    integer :: stat
+    real :: current_cpu_time, current_wall_time
+    character(len=OPTION_PATH_LEN) :: func
+
+    should_output = .false.
+
+    if (.not. CS%last_times_initialised) then
+      ! if the last_dump*_time variables have not been initialised, assume there should be output
+
+      should_output = .true.
+      return
+    end if
+
+    if (have_option(trim(option_path) // "/dump_period")) then
+      ! dump period isn't set, or dump period has elapsed since the last time there was a dump
+
+      if (CS%real_dump_period == 0.0 .or. dump_count_greater(current_time, CS%last_dump_time, CS%real_dump_period)) then
+        if (have_option(trim(option_path) // "/dump_period/constant")) then
+          call get_option(trim(option_path) // "/dump_period/constant", CS%real_dump_period)
+        else if (have_option(trim(option_path) // "/dump_period/python")) then
+          call get_option(trim(option_path) // "/dump_period/python", func)
+          call real_from_python(func, current_time, CS%real_dump_period)
+        else
+          FLAbort("Unable to determine dump period type.")
+        end if
+
+        if (CS%real_dump_period < 0.0) then
+          FLExit("Dump period cannot be negative.")
+        end if
+
+        should_output = .true.
+        return
+      end if
+    end if
+
+    if (have_option(trim(option_path) // "/dump_period_in_timesteps")) then
+      ! timestep dump period isn't set, or the required number of timesteps has passed since the last dump
+
+      if (CS%int_dump_period == 0 .or. mod(timestep, CS%int_dump_period) == 0) then
```

Comment:
I've seen this pattern in a few places (e.g. the `dump_count_greater` checks above, and in at least [Write_State.F90](https://github.com/FluidityProject/fluidity/blob/master/femtools/Write_State.F90#L73-L122) and [qmesh.F90](https://github.com/FluidityProject/fluidity/blob/master/assemble/qmesh.F90#L90-L108)), and I think it's strictly incorrect: the Fortran compiler isn't required to short-circuit the evaluation here if `int_dump_period = 0`, in which case we get a divide by zero. In practice, gfortran will perform this optimisation above `-O0` or with `-ffrontend-optimize`, but it did actually just bite me on a debug build where this is disabled.

I realised that I've replicated the logic here from Write_State. I don't know if it's worth fixing all these occurrences, but it's at least something to keep in mind for certain failures.

Reply:
Yeah, I'm aware - it's a hard to beat habit it seems. I'd say don't introduce any new cases and fix existing cases if you feel like it. It also tends to add valgrind errors